packages/repository/README.md

```diff
@@ -21,47 +21,137 @@ $ npm install --save @loopback/repository
 
 ## Basic use
 
+At the moment, we only have implementations of `Repository` based on LoopBack
+3.x `loopback-datasource-juggler` and connectors. The following steps illustrate
+how to define repositories and use them with controllers.
+
+### Defining a legacy datasource and a model
+
 The repository module provides APIs to define LoopBack 3.x data sources and
 models. For example,
 
 ```ts
-import {
-  DataSourceConstructor,
-  juggler,
-  Entity,
-  model,
-  ModelDefinition
-} from '@loopback/repository';
+// src/datasources/db.datasource.ts
+import {juggler, DataSourceConstructor} from '@loopback/repository';
 
-export const ds: juggler.DataSource = new DataSourceConstructor({
+export const db: juggler.DataSource = new DataSourceConstructor({
   name: 'db',
   connector: 'memory',
 });
+```
+
+```ts
+// src/models/note.model.ts
+import {model, Entity, property} from '@loopback/repository';
 
 @model()
 export class Note extends Entity {
-  static definition = new ModelDefinition({
-    name: 'note',
-    properties: {
-      id: {name: 'id', type: 'number', id: true},
-      title: 'string',
-      content: 'string'
-    }
-  })
-};
+  @property({id: true})
+  id: string;
+  @property() title: string;
+  @property() content: string;
+}
+```
+
+**NOTE**: There is no declarative support for data source and model yet in
+LoopBack 4. These constructs need to be created programmatically as
+illustrated above.
+
+### Defining a repository
+
+A repository can be created by extending `DefaultCrudRepository` and using
+dependency injection to resolve the datasource.
+
+```ts
+// src/repositories/note.repository.ts
+import {DefaultCrudRepository, DataSourceType} from '@loopback/repository';
+import {Note} from '../models';
+import {inject} from '@loopback/core';
+
+export class NoteRepository extends DefaultCrudRepository<
+  Note,
+  typeof Note.prototype.id
+> {
+  constructor(@inject('datasources.db') protected dataSource: DataSourceType) {
+    super(Note, dataSource);
+  }
+}
 ```
-A repository can be created directly using `DefaultCrudRepository`.
+
+### Defining a controller
+
+Controllers serve as handlers for API requests. We declare controllers as
+classes with optional dependency injection by decorating constructor parameters
+or properties.
 
 ```ts
-import {DefaultCrudRepository} from '@loopback/repository';
-// also import Note and ds from wherever you defined them
+// src/controllers/note.controller.ts
+import {repository} from '@loopback/repository';
+import {NoteRepository} from '../repositories';
+import {Note} from '../models';
+import {post, requestBody, get, param} from '@loopback/openapi-v3';
+
+export class NoteController {
+  constructor(
+    // Use constructor dependency injection to set up the repository
+    @repository(NoteRepository.name) public noteRepo: NoteRepository,
+  ) {}
+
+  // Create a new note
+  @post('/note')
+  create(@requestBody() data: Note) {
```

Comment:
I think `data` should be `Partial<Note>`? E.g. when the database is creating the `id` value for us, then `id` needs to be omitted from the data payload.

Feel free to leave this change out of scope of this pull request though.

Reply:
I just tried out `Partial<Note>` and see what OpenAPI spec it would generate, and it just interpreted the parameter type to be an `object`, so it seems like we can't do this here unless we specifically circumvent around this issue.

This isn't really a problem right now since type validation isn't being done, but it IS technically wrong for the route to accept an object that is only partially a `Note`. What do you think?

Reply:
@bajtos ^^^

Reply:
@shimks I'm interested to hear why the route can't accept `Partial<Note>`. I thought partial just meant that the type is preserved, but some(or all?) properties are optional.

Reply:
Yes, `Partial<Note>` means `Note` model with all properties made optional. We are already using it in our ORM layer, see e.g. here:

https://github.com/strongloop/loopback-next/blob/1d9979fe4a7d0ef828cd9972694b49be283d7ca8/packages/repository/src/legacy-juggler-bridge.ts#L110-L113

> I just tried out changing the type to `Partial<Note>` to see what OpenAPI spec would get generated, and it just interpreted the parameter type to be an object

That looks like a rather unfortunate limitation of what can be inferred via TypeScript type system 😞 

I personally think it's important to have correct type annotation (`Partial<Note>`), even if it means that we have to explicitly provide the request body schema for OpenAPI Spec.